File: PowerCalculation.m
clc;
clear;

I_origin = csvread('data/ALL0001/OUTPUT.CSV', 0, 1, [0, 1, 2499, 1]);
U_origin = csvread('data/ALL0001/OUTPUT.CSV', 0, 0, [0, 0, 2499, 0]);

t = linspace(0,2500,2500);

U = smooth(U_origin,30,'rlowess');
I = smooth(I_origin,30,'rlowess');  		% 利用rlowess方法对y进行平滑处理

figure;  							        % 新建一个图形窗口

plot(t,I_origin,'k:');  			        % 绘制加噪波形图
hold on;      
plot(t,I,'k','linewidth',3);  	        % 绘制平滑后波形图
xlabel('t');  					        % 为X轴加标签
ylabel('rlowess');  				        % 为Y轴加标签
legend('加噪波形','平滑后波形');

power_origin = zeros(2500, 1);
power_handled = zeros(2500, 1);
total_power_origin = 0;
total_power_handled = 0;
for i = 1:2500
    power_origin(i, 1) = U_origin(i, 1) * I_origin(i, 1);
    power_handled(i, 1) = U(i, 1) * I(i, 1);
    total_power_origin = total_power_origin + power_origin(i, 1);
    total_power_handled = total_power_handled + power_handled(i, 1);
end

sample_interval = 2e-8;
total_power_handled = total_power_handled * sample_interval;
total_power_origin = total_power_origin * sample_interval;

File: powerCalculation.py
# %%
# coding=utf-8

from cProfile import label
from encodings import utf_8
import numpy as np
import csv

path_current = './data/ALL0000/F0000CH2.CSV'
path_voltage = './data/ALL0000/F0000CH1.CSV'
with open(path_voltage, mode='r') as f:
    data_current = np.loadtxt(f, str, delimiter=',')
    y_current_origin = data_current[:, 4]
    y_current_origin = [np.float64(i) for i in y_current_origin]
    # print(data[:, 4])
    f.close()
with open(path_current, mode='r') as f:
    data_voltage = np.loadtxt(f, str, delimiter=',')
    y_voltage_origin = data_voltage[:, 4]
    y_voltage_origin = [np.float64(i) for i in y_voltage_origin]
    # print(data[:, 4])
    f.close()

# %%
import statsmodels.api as sm 
# pip install -i https://pypi.tuna.tsinghua.edu.cn/simple scipy==1.2.1 --upgrade
lowess = sm.nonparametric.lowess
import matplotlib.pyplot as plt

n = 2500
x = np.linspace(0, 2500, n)
# yest = lowess(y, x)
y_current_handled = lowess(y_current_origin, x, frac=0.015)[:,1] # 选0.015
y_voltage_handled = lowess(y_voltage_origin, x, frac=0.015)[:,1] 

# %%
plt.figure()
ax1 = plt.subplot()
ax2 = ax1.twinx()
ax1.plot(x, y_voltage_handled, color='orange', label='voltage')
ax2.plot(x, y_current_handled, label='current')

ax1.set_ylabel('voltage', color='orange')
ax2.set_ylabel('current', color='b')
plt.legend()
plt.show()


# %%
power_origin = np.array([], dtype=np.float64)
for i in range(2500):
    temp_power = y_voltage_origin[i] * y_current_origin[i]
    power_origin = np.append(power_origin, temp_power)

# %%
plt.figure()
plt.plot(x, power_origin)
plt.legend()
plt.show()

# %%
power_handled = np.array([], dtype=np.float64)
for i in range(2500):
    temp_power = y_voltage_handled[i] * y_current_handled[i]
    power_handled = np.append(power_handled, temp_power)

# %%
plt.figure()
plt.plot(x, power_handled)
plt.legend()
plt.show()

# %%
# 离散功率求和
total_power_origin = 0
total_power_handled = 0
for i in range(2500):
    total_power_origin += power_origin[i]
    total_power_handled += power_handled[i]

# print("total power (handled) is %f" % total_power_handled)
# print("total power (origin) is %f" % total_power_origin)

# %%
# 功率换算（乘以单位时间）

sample_interval = 2e-8 # 本例中最小时间尺度为20ns
total_power_origin *= sample_interval
total_power_handled *= sample_interval

print("total power (handled) is %f" % total_power_handled)
print("total power (origin) is %f" % total_power_origin)

# %%
